### PR TITLE
8292741: Convert JvmtiTagMapTable to ResourceHashtable

### PR DESCRIPTION
**Pull request by afshin-zafari**
test of tier1-5 passed.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8292741](https://bugs.openjdk.org/browse/JDK-8292741): Convert JvmtiTagMapTable to ResourceHashtable

### Reviewers
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**)
 * [Coleen Phillimore](https://openjdk.org/census#coleenp) (@coleenp - **Reviewer**) ⚠️ Review applies to [b5a93bd5](https://git.openjdk.org/jdk/pull/11288/files/b5a93bd552445e19274d5ab8ff1041c963804c81)
 * [Robbin Ehn](https://openjdk.org/census#rehn) (@robehn - **Reviewer**) ⚠️ Review applies to [d55d7612](https://git.openjdk.org/jdk/pull/11288/files/d55d761229ce8496eb448234011a4fe844a2b553)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/11288/head:pull/11288` \
`$ git checkout pull/11288`

Update a local copy of the PR: \
`$ git checkout pull/11288` \
`$ git pull https://git.openjdk.org/jdk pull/11288/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 11288`

View PR using the GUI difftool: \
`$ git pr show -t 11288`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/11288.diff">https://git.openjdk.org/jdk/pull/11288.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/share/prims/jvmtiTagMap.cpp [hunk 1]
```diff
@@ -170,23 +170,7 @@ void JvmtiTagMap::check_hashmaps_for_heapwalk(GrowableArray<jlong>* objects) {
...
-}
-
-
-// A CallbackWrapper is a support class for querying and tagging an object
-// around a callback to a profiler. The constructor does pre-callback
-// work to get the tag value, klass tag value, ... and the destructor
-// does the post-callback work of tagging or untagging the object.
-//
-// {
+  return tag_map->hashmap()->find(o); } // A CallbackWrapper is a support class for querying and tagging an object // around a callback to a profiler. The constructor does pre-callback // work to get the tag value, klass tag value, ... and the destructor // does the post-callback work of tagging or untagging the object. // // {
```
Comment (dholmes-ora) [inline src/hotspot/share/prims/jvmtiTagMap.cpp:173]: The formatting of this line is messed up.

File: src/hotspot/share/prims/jvmtiTagMap.cpp [hunk 2]
```diff
@@ -225,10 +209,9 @@ class CallbackWrapper : public StackObj {
     // record the context
     _tag_map = tag_map;
     _hashmap = tag_map->hashmap();
-    _entry = _hashmap->find(_o);
+    _obj_tag  = _hashmap->find(_o);
 
     // get object tag
-    _obj_tag = (_entry == NULL) ? 0 : _entry->tag();
```
Comment (dholmes-ora) [inline src/hotspot/share/prims/jvmtiTagMap.cpp:215]: The comment and action are now out-of-order.

File: src/hotspot/share/prims/jvmtiTagMap.cpp [hunk 3]
```diff
@@ -265,8 +249,9 @@ void inline CallbackWrapper::post_callback_tag_update(oop o,
     if (obj_tag == 0) {
       hashmap->remove(o);
     } else {
-      if (obj_tag != entry->tag()) {
-         entry->set_tag(obj_tag);
+      if (obj_tag != current_tag ) {
+        hashmap->remove(o);
+        hashmap->add(o, obj_tag);
```
Comment (dholmes-ora) [inline src/hotspot/share/prims/jvmtiTagMap.cpp:254]: This change is not atomic - is that a problem? The concurrency aspects of using this map are not clear.

File: src/hotspot/share/prims/jvmtiTagMap.cpp [hunk 4]
```diff
@@ -313,10 +298,9 @@ class TwoOopCallbackWrapper : public CallbackWrapper {
       _referrer = referrer;
       // record the context
       _referrer_hashmap = tag_map->hashmap();
-      _referrer_entry = _referrer_hashmap->find(_referrer);
+      _referrer_obj_tag  = _referrer_hashmap->find(_referrer);
 
       // get object tag
```
Comment (dholmes-ora) [inline src/hotspot/share/prims/jvmtiTagMap.cpp:303]: Action the comment and action are now out-of-order.

File: src/hotspot/share/prims/jvmtiTagMap.cpp [hunk 5]
```diff
@@ -326,9 +310,9 @@ class TwoOopCallbackWrapper : public CallbackWrapper {
 
   ~TwoOopCallbackWrapper() {
     if (!is_reference_to_self()){
+
```
Comment (dholmes-ora) [inline src/hotspot/share/prims/jvmtiTagMap.cpp:313]: Unnecessary blank line added.

File: src/hotspot/share/prims/jvmtiTagMap.cpp [hunk 6]
```diff
@@ -357,12 +341,14 @@ void JvmtiTagMap::set_tag(jobject object, jlong tag) {
   // resolve the object
   oop o = JNIHandles::resolve_non_null(object);
 
-  // see if the object is already tagged
+  //JvmtiTagMapEntry entry ;
+  //_hashmap->add_update_remove(&entry, o, tag);
```
Comment (dholmes-ora) [inline src/hotspot/share/prims/jvmtiTagMap.cpp:345]: What are these comments for?

File: src/hotspot/share/prims/jvmtiTagMap.cpp [hunk 7]
```diff
@@ -375,7 +361,8 @@ void JvmtiTagMap::set_tag(jobject object, jlong tag) {
     if (tag == 0) {
       hashmap->remove(o);
     } else {
-      entry->set_tag(tag);
+        hashmap->remove(o);
+        hashmap->add(o,tag);
```
Comment (dholmes-ora) [inline src/hotspot/share/prims/jvmtiTagMap.cpp:365]: Nit: need space after comma

File: src/hotspot/share/prims/jvmtiTagMap.cpp [hunk 8]
```diff
@@ -1271,25 +1258,26 @@ class TagObjectCollector : public JvmtiTagMapEntryClosure {
   // - if it matches then we create a JNI local reference to the object
   // and record the reference and tag value.
   //
-  void do_entry(JvmtiTagMapEntry* entry) {
+  bool do_entry(JvmtiTagMapEntry & key , jlong & value ) {
```
Comment (dholmes-ora) [inline src/hotspot/share/prims/jvmtiTagMap.cpp:1261]: Nit: no space before &

File: src/hotspot/share/prims/jvmtiTagMap.cpp [hunk 9]
```diff
@@ -1271,25 +1258,26 @@ class TagObjectCollector : public JvmtiTagMapEntryClosure {
...
     for (int i=0; i<_tag_count; i++) {
```
Comment (dholmes-ora) [inline src/hotspot/share/prims/jvmtiTagMap.cpp:1262]: Pre-existing: need spaces around binary operators = and <

File: src/hotspot/share/prims/jvmtiTagMapTable.cpp [hunk 10]
```diff
@@ -37,190 +37,103 @@
...
-        ResourceMark rm;
-        log_trace(jvmti, table)("JvmtiTagMap entry found for %s index %d",
-                                obj->print_value_string(), index);
-        // The object() accessor makes sure the target object is kept alive before
-        // leaking out.
-        (void)p->object();
-        return p;
-      }
+void JvmtiTagMapTable::clear() {
+  struct RemoveAll{
```
Comment (dholmes-ora) [inline src/hotspot/share/prims/jvmtiTagMapTable.cpp:77]: Nit: space before {

Stylistically I'm not sure we define local structs like this when defining a "closure" object. ??

File: src/hotspot/share/prims/jvmtiTagMapTable.cpp [hunk 11]
```diff
@@ -37,190 +37,103 @@
...
-  JvmtiTagMapEntry* p = new_entry(hash, w, tag);
-  Hashtable<WeakHandle, mtServiceability>::add_entry(index, p);
-  ResourceMark rm;
-  log_trace(jvmti, table)("JvmtiTagMap entry added for %s index %d",
-                          obj->print_value_string(), index);
-
-  // Resize if the table is getting too big.
-  resize_if_needed();
+JvmtiTagMapTable::~JvmtiTagMapTable() {
+  clear();
```
Comment (dholmes-ora) [inline src/hotspot/share/prims/jvmtiTagMapTable.cpp:89]: Is this the only use of `clear`? If so we can just inline its code here and remove it from the API.

File: src/hotspot/share/prims/jvmtiTagMapTable.cpp [hunk 12]
```diff
@@ -37,190 +37,103 @@
...
+}
 
-  return p;
+jlong JvmtiTagMapTable::find(oop obj) {
+   //if (obj->fast_no_hash_check()) {
+   //  return 0;
+   //} else {
```
Comment (dholmes-ora) [inline src/hotspot/share/prims/jvmtiTagMapTable.cpp:95]: What are these comments?

File: src/hotspot/share/prims/jvmtiTagMapTable.cpp [hunk 13]
```diff
@@ -37,190 +37,103 @@
...
+     JvmtiTagMapEntry jtme(obj);
+     jlong* found = _rrht_table.get(jtme);
+     return found == NULL ? 0 : *found;
+   //}
+ }
+
+bool JvmtiTagMapTable::add(oop obj, jlong tag) {
```
Comment (dholmes-ora) [inline src/hotspot/share/prims/jvmtiTagMapTable.cpp:102]: I'm not seeing that a return value has any use here when it is always expected to be true.

File: src/hotspot/share/prims/jvmtiTagMapTable.cpp [hunk 14]
```diff
@@ -37,190 +37,103 @@
...
-        oops_removed++;
-        log_trace(jvmti, table)("JvmtiTagMap entry removed for index %d", i);
-        jlong tag = entry->tag();
-        *p = entry->next();
-        free_entry(entry);
-
-        // collect object tags for posting JVMTI events later
-        if (objects != NULL) {
-          objects->append(tag);
+  struct IsDead{
```
Comment (dholmes-ora) [inline src/hotspot/share/prims/jvmtiTagMapTable.cpp:123]: Nit: space before {

Same query about using a local struct for this.

File: src/hotspot/share/prims/jvmtiTagMapTable.cpp [hunk 15]
```diff
@@ -37,190 +37,103 @@
...
+    GrowableArray<jlong>* _objects;
+    IsDead(GrowableArray<jlong>* objects) : _objects(objects){}
+    bool do_entry(JvmtiTagMapEntry const & entry, jlong tag){
+      if ( entry.object_no_keepalive() == NULL){
+        if(_objects!=NULL){
```
Comment (dholmes-ora) [inline src/hotspot/share/prims/jvmtiTagMapTable.cpp:128]: Nit: need space before {
Nit: need space after if
Nit: no space after (

File: src/hotspot/share/prims/jvmtiTagMapTable.hpp [hunk 16]
```diff
@@ -25,67 +25,73 @@
...
+};
 
-  JvmtiTagMapEntry* new_entry(unsigned int hash, WeakHandle w, jlong tag);
-  void free_entry(JvmtiTagMapEntry* entry);
+  typedef
+  ResizeableResourceHashtable <JvmtiTagMapEntry, jlong,
+                               AnyObj::C_HEAP, mtInternal,
+                               JvmtiTagMapEntry::get_hash,
+                               JvmtiTagMapEntry::equals
+                               > ResizableResourceHT ;
```
Comment (dholmes-ora) [inline src/hotspot/share/prims/jvmtiTagMapTable.hpp:69]: Nit: keep > on previous line
Nit: no space before ;

File: src/hotspot/share/prims/jvmtiTagMapTable.hpp [hunk 17]
```diff
@@ -25,67 +25,73 @@
...
+class JvmtiTagMapTable : public CHeapObj<mtInternal> {
+  enum Constants {
+    _table_size  = 1007
+  };
 
-  JvmtiTagMapEntry* find(int index, unsigned int hash, oop obj);
+private:
 
   void resize_if_needed();
+  ResizableResourceHT  _rrht_table;
```
Comment (dholmes-ora) [inline src/hotspot/share/prims/jvmtiTagMapTable.hpp:79]: This can just be `_table` - no need for the `rrht` prefix.

File: src/hotspot/share/prims/jvmtiTagMapTable.hpp [hunk 18]
```diff
@@ -94,8 +100,8 @@ class JvmtiTagMapTable : public Hashtable<WeakHandle, mtServiceability> {
 
 // A supporting class for iterating over all entries in Hashmap
 class JvmtiTagMapEntryClosure {
- public:
-  virtual void do_entry(JvmtiTagMapEntry* entry) = 0;
+  public:
+  virtual bool do_entry(JvmtiTagMapEntry & key , jlong & value) = 0;
```
Comment (dholmes-ora) [inline src/hotspot/share/prims/jvmtiTagMapTable.hpp:104]: Nit: no space needed before &

**Review by dholmes-ora — CHANGES_REQUESTED** [review]
Hi Afshin,

The general conversion approach seems okay, though it is hard to see exactly how the new table is used compared to the old table  i.e. in relation to the role of `JvmtiTagMapEntry` now that it is not needed for the actual table. ??

A few queries below and a lot of nits to fix up (sorry).

Thanks.

File: src/hotspot/share/prims/jvmtiTagMap.cpp (hunk 3 above)
Comment (afshin-zafari) [reply src/hotspot/share/prims/jvmtiTagMap.cpp:254]: add() is supposed to add a new object with tag. There is an assert() in its body that checks it. By removing the assert, add() can be used for updating as well.

File: src/hotspot/share/prims/jvmtiTagMapTable.cpp (hunk 10 above)
Comment (afshin-zafari) [reply src/hotspot/share/prims/jvmtiTagMapTable.cpp:77]: I have seen these local structs in some places in the code (e.g., in concurrentHashTable.hpp), therefore I used it this way. 
Any preference?

File: src/hotspot/share/prims/jvmtiTagMapTable.cpp (hunk 12 above)
Comment (afshin-zafari) [reply src/hotspot/share/prims/jvmtiTagMapTable.cpp:95]: Coleen's suggestion for efficiency reasons.

File: src/hotspot/share/prims/jvmtiTagMapTable.cpp (hunk 13 above)
Comment (afshin-zafari) [reply src/hotspot/share/prims/jvmtiTagMapTable.cpp:102]: ResourceHashTable::put() returns true if the Key,Value is added, false if the Value is updated.

File: src/hotspot/share/prims/jvmtiTagMapTable.cpp (hunk 14 above)
Comment (afshin-zafari) [reply src/hotspot/share/prims/jvmtiTagMapTable.cpp:123]: Alternative for struct?

Comment (dholmes-ora) [reply src/hotspot/share/prims/jvmtiTagMapTable.cpp:123]: Not sure ... didn't we start using C++ lambda's for some of these "closure" operations? @coleenp what is the usual pattern we use for this kind of thing?

File: src/hotspot/share/prims/jvmtiTagMap.cpp (hunk 3 above)
Comment (dholmes-ora) [reply src/hotspot/share/prims/jvmtiTagMap.cpp:254]: Are you suggesting that `add`  can also act as a `replace` operation? I would think we would want a separate method for that.

File: src/hotspot/share/prims/jvmtiTagMapTable.cpp (hunk 12 above)
Comment (dholmes-ora) [reply src/hotspot/share/prims/jvmtiTagMapTable.cpp:95]: If the comment is meant to suggest some possible future optimisation it should say that.

File: src/hotspot/share/prims/jvmtiTagMapTable.cpp (hunk 13 above)
Comment (dholmes-ora) [reply src/hotspot/share/prims/jvmtiTagMapTable.cpp:102]: But this doesn't do that, so ??

File: src/hotspot/share/prims/jvmtiTagMapTable.cpp (hunk 14 above)
Comment (afshin-zafari) [reply src/hotspot/share/prims/jvmtiTagMapTable.cpp:123]: The `unlink` method of `ResourceHashTable` gets an ITER type and calls its `do_entry(Key&,Value&)` method. 
If we want to use lambdas, `unlink` should call the input directly and not one of its methods.

File: src/hotspot/share/prims/jvmtiTagMapTable.cpp (hunk 12 above)
Comment (coleenp) [reply src/hotspot/share/prims/jvmtiTagMapTable.cpp:95]: Oh you should uncomment that code. The hashtable that this replaces has the same early cut-out for efficiency.   You can also have an early cut out if the hashtable is empty, like if (is_empty()) { return 0 }

File: src/hotspot/share/prims/jvmtiTagMapTable.cpp (hunk 14 above)
Comment (coleenp) [reply src/hotspot/share/prims/jvmtiTagMapTable.cpp:123]: We don't have a lambda version of unlink (yet).  The local struct is fine IMO.

File: src/hotspot/share/prims/jvmtiTagMapTable.hpp [hunk 19]
```diff
@@ -25,67 +25,73 @@
...
-
-// Hashtable to record oops used for JvmtiTagMap
 class JvmtiTagMapEntryClosure;
 
-class JvmtiTagMapEntry : public HashtableEntry<WeakHandle, mtServiceability> {
-  jlong _tag;                           // the tag
- public:
-  JvmtiTagMapEntry* next() const {
-    return (JvmtiTagMapEntry*)HashtableEntry<WeakHandle, mtServiceability>::next();
-  }
```
Comment (coleenp) [inline src/hotspot/share/prims/jvmtiTagMapTable.hpp:41]: This should add a comment above JvmtiTagMapEntry that the entry is the key (could be renamed JvmtiTagMapKey).  Also add a comment above that the oop is needed for lookup rather than creating a WeakHandle during lookup because the HeapWalker may walk soon to be dead objects and creating a WeakHandle for an otherwise dead object makes G1 unhappy.

File: src/hotspot/share/prims/jvmtiTagMap.cpp (hunk 3 above)
Comment (afshin-zafari) [reply src/hotspot/share/prims/jvmtiTagMap.cpp:254]: I didn't notice the suggested name (`replace()`), so It is implemented as `update()` method.

**Comment by dholmes-ora** [comment]
This PR seems to be broken now - full of unexpected changes.

**Comment by afshin-zafari** [comment]
> This PR seems to be broken now - full of unexpected changes.

The PR changed files are correct now.

File: src/hotspot/share/prims/jvmtiTagMap.cpp [hunk 20]
```diff
@@ -1271,25 +1259,26 @@ class TagObjectCollector : public JvmtiTagMapEntryClosure {
   // - if it matches then we create a JNI local reference to the object
   // and record the reference and tag value.
   //
-  void do_entry(JvmtiTagMapEntry* entry) {
-    for (int i=0; i<_tag_count; i++) {
-      if (_tags[i] == entry->tag()) {
+  bool do_entry(JvmtiTagMapKey& key, jlong& value) {
```
Comment (dholmes-ora) [inline src/hotspot/share/prims/jvmtiTagMap.cpp:1262]: I still see no point in this method returning a bool when it only ever returns true.

File: src/hotspot/share/prims/jvmtiTagMapTable.cpp (hunk 11 above)
Comment (dholmes-ora) [reply src/hotspot/share/prims/jvmtiTagMapTable.cpp:86]: This issue is not resolved.

File: src/hotspot/share/prims/jvmtiTagMapTable.cpp (hunk 13 above)
Comment (dholmes-ora) [reply src/hotspot/share/prims/jvmtiTagMapTable.cpp:103]: This issue is not resolved.

File: src/hotspot/share/prims/jvmtiTagMapTable.cpp [hunk 21]
```diff
@@ -38,189 +38,109 @@
...
-      return; // done
-    }
-    // get next entry and address
-    p = entry->next_addr();
-    entry = entry->next();
-  }
+bool JvmtiTagMapTable::update(oop obj, jlong tag) {
+  JvmtiTagMapKey new_entry(obj);
+  bool is_updated = _table.put(new_entry, tag) == false;
+  assert(is_updated, "should be updated and not added");
```
Comment (dholmes-ora) [inline src/hotspot/share/prims/jvmtiTagMapTable.cpp:114]: This API is confusing. You have `add` and `replace` but in product mode an `add` can `replace` and a `replace` can `add`. It isn't clear if these should be allowed rare conditions or fatal errors.

**Review by dholmes-ora — COMMENTED** [review]
I'm still seeing a few oddities in the code. Overall it is very hard in the PR to get a good sense of the conversion process, but it seems fairly reasonable. It is hard to see how all the different parts of the APIs connect e.g. to understand why some methods are returning boolean values that are always true.

Also please do not force-push.

**Comment by afshin-zafari** [comment]
> > Also please do not force-push.
> 
> FYI: I'm not sure it even was possible to fix the "full of unexpected changes" without force push. Even if it was, the effort required to fix it without force push would be pretty significant. So I'd say this one of those exceptions to the rule :)

I did 2 mistakes: rebase my branch while a PR was in progress, and force-push my changes
After getting help from colleagues, I had to do another force-push to get back my branch to its state before the mistakes.

File: src/hotspot/share/prims/jvmtiTagMapTable.hpp [hunk 22]
```diff
@@ -25,77 +25,87 @@
...
   }
+};
 
-  JvmtiTagMapEntry* new_entry(unsigned int hash, WeakHandle w, jlong tag);
-  void free_entry(JvmtiTagMapEntry* entry);
-
-  unsigned int compute_hash(oop obj);
+typedef
+ResizeableResourceHashtable <JvmtiTagMapKey, jlong,
+                              AnyObj::C_HEAP, mtInternal,
```
Comment (coleenp) [inline src/hotspot/share/prims/jvmtiTagMapTable.hpp:72]: This should be mtServiceability rather than mtInternal.

**Review by coleenp — APPROVED** [review]
I have a couple of minor comments but this looks good.  This hashtable was more problematic to convert so good job!

**Comment by dholmes-ora** [comment]
The alternative to force-push would have been just to create a new PR.

File: src/hotspot/share/prims/jvmtiTagMapTable.cpp [hunk 23]
```diff
@@ -38,189 +38,109 @@
...
+jlong JvmtiTagMapTable::find(oop obj) {
+  if (is_empty()) {
+    return 0;
   }
-  return NULL;
-}
 
-JvmtiTagMapEntry* JvmtiTagMapTable::find(oop obj) {
   if (obj->fast_no_hash_check()) {
-    // Objects in the table all have a hashcode.
```
[the remark this answers is not in the document]
Comment (afshin-zafari) [reply src/hotspot/share/prims/jvmtiTagMapTable.cpp:117]: Done

File: src/hotspot/share/prims/jvmtiTagMap.cpp (hunk 20 above)
Comment (afshin-zafari) [reply src/hotspot/share/prims/jvmtiTagMap.cpp:1262]: In the `iterate` method of ResourceHashTable, in respurceHash.hpp lines 227-240 copied here, requires a function that returns boolean. If returned false the iterate will break.
```
 template<typename Function>
  void iterate(Function function) const { // lambda enabled API
    Node* const* bucket = table();
    const unsigned sz = table_size();
    while (bucket < bucket_at(sz)) {
      Node* node = *bucket;
      while (node != NULL) {
        bool cont = function(node->_key, node->_value);    // <--------------******
        if (!cont) { return; }
        node = node->_next;
      }
      ++bucket;
    }
  }
```
The other `iterate` methods are wrappers around this one.
Always returning true means to continue iterating over all the existing items. 
The former base table for jvmtiTagMapTable needs the `do_entry` be `void`.

File: src/hotspot/share/prims/jvmtiTagMapTable.cpp (hunk 21 above)
Comment (afshin-zafari) [reply src/hotspot/share/prims/jvmtiTagMapTable.cpp:114]: I agree that this is ambigious. The `jvmtiTagMap` calls `add/update` methods of `jvmtiTagMapTable` which in turn calls `resourceHashTable` methods `put` and `put_if_absent`.
`put` and `put_if_absent` can be used for both adding and updating and reporting it in returned values.
`jvmtiTagMap` calls to `jvmtiTagMapTable` add/update  DO NOT CARE about the returned values. For these calls, it doesn't mater if an add (update) method resulted in an update (add). 
So which one of the following cases would be correct?
 - Based on the `jvmtiTagMap` calls, let the add/update be void and ignore the differences of what really happens in the table. Or
 - Based on the `resourceHashTable` bool-valued `put` and `put_if_absent` methods, let the add/update be bool-values and leave the decision on "what really happened in the table" to the callers in `jvmtiTagMap`. (current implementation)

File: src/hotspot/share/prims/jvmtiTagMap.cpp (hunk 20 above)
Comment (dholmes-ora) [reply src/hotspot/share/prims/jvmtiTagMap.cpp:1262]: Okay in that case please add to the comment preceding this method:
```
// Always return true so the iteration continues.
```

File: src/hotspot/share/prims/jvmtiTagMapTable.cpp (hunk 21 above)
Comment (dholmes-ora) [reply src/hotspot/share/prims/jvmtiTagMapTable.cpp:114]: The issue is not the underlying RHT methods but the semantics of the `jvmtiTagMap` methods. If a call to add always expects to add then it should be a fatal error if it actually did an update as that indicates something is broken. Similarly if an update actually does an add.

**Review by robehn — APPROVED** [review]
Looks fine, thanks. (+- DavidH comment)

File: src/hotspot/share/prims/jvmtiTagMapTable.cpp (hunk 21 above)
Comment (coleenp) [reply src/hotspot/share/prims/jvmtiTagMapTable.cpp:114]: The JvmtiTagMap code adds/updates and removes the entries like below in two places which could probably be simplified, but I think that's outside the scope of this RFE.   I suggest removing the bool return from add, remove and update, and add the assert(true) in the remove case.  There's already an assert that add/update happened in the other cases.

 ``` // if the object is not already tagged then we tag it
  if (found_tag == 0) {
    if (tag != 0) {
      hashmap->add(o, tag);
    } else {
      // no-op
    }
  } else {
    // if the object is already tagged then we either update
    // the tag (if a new tag value has been provided)
    // or remove the object if the new tag value is 0.
    if (tag == 0) {
      hashmap->remove(o);
    } else {
      hashmap->update(o, tag);
    }
  }
```

File: src/hotspot/share/prims/jvmtiTagMap.cpp [hunk 24]
```diff
@@ -191,26 +183,27 @@ static inline jlong tag_for(JvmtiTagMap* tag_map, oop o) {
...
+// wrapper goes out of scope here which results in the destructor
+// checking to see if the object has been tagged, untagged, or the
+// tag value has changed.
 //
 class CallbackWrapper : public StackObj {
  private:
   JvmtiTagMap* _tag_map;
   JvmtiTagMapTable* _hashmap;
-  JvmtiTagMapEntry* _entry;
+  bool _entry_found;
```
Comment (coleenp) [inline src/hotspot/share/prims/jvmtiTagMap.cpp:195]: is this used?

File: src/hotspot/share/prims/jvmtiTagMapTable.hpp [hunk 25]
```diff
@@ -25,77 +25,87 @@
...
+class JvmtiTagMapKeyClosure;
+
+// The oop is needed for lookup rather than creating a WeakHandle during
+// lookup because the HeapWalker may walk soon to be dead objects and
+// creating a WeakHandle for an otherwise dead object makes G1 unhappy.
+//
+// This class is the Key type for inserting in ResizeableResourceHashTable
+// Its get_hash() and equals() methods are also used for getting the hash
+// value of a Key and comparing two Keys, respectively.
+class JvmtiTagMapKey : public CHeapObj<mtInternal> {
```
Comment (coleenp) [inline src/hotspot/share/prims/jvmtiTagMapTable.hpp:44]: Can you change mtInternal to mtServiceability here?

**Review by coleenp — CHANGES_REQUESTED** [review]

File: src/hotspot/share/prims/jvmtiTagMapTable.hpp [hunk 26]
```diff
@@ -25,77 +25,87 @@
...
-
-  unsigned int compute_hash(oop obj);
+typedef
+ResizeableResourceHashtable <JvmtiTagMapKey, jlong,
+                              AnyObj::C_HEAP, mtServiceability,
+                              JvmtiTagMapKey::get_hash,
+                              JvmtiTagMapKey::equals> ResizableResourceHT;
 
-  JvmtiTagMapEntry* find(int index, unsigned int hash, oop obj);
+class JvmtiTagMapTable : public CHeapObj<mtInternal> {
```
Comment (coleenp) [inline src/hotspot/share/prims/jvmtiTagMapTable.hpp:76]: Change this to mtServiceability

File: src/hotspot/share/prims/jvmtiTagMap.cpp [hunk 27]
```diff
@@ -447,7 +436,7 @@ ClassFieldMap::ClassFieldMap() {
 }
 
 ClassFieldMap::~ClassFieldMap() {
-  for (int i=0; i<_fields->length(); i++) {
+  for (int i = 0; i < _fields->length(); i++) {
     delete _fields->at(i);
   }
   delete _fields;
```
Comment (coleenp) [inline src/hotspot/share/prims/jvmtiTagMap.cpp:442]: The changes here to ClassFieldMap are not part of the change, can you revert them (also leave mtInternal even though that's wrong).

File: src/hotspot/share/prims/jvmtiTagMap.hpp [hunk 28]
```diff
@@ -32,7 +32,7 @@
 
 class JvmtiEnv;
 class JvmtiTagMapTable;
-class JvmtiTagMapEntryClosure;
+class JvmtiTagMapKeyClosure;
 
 class JvmtiTagMap :  public CHeapObj<mtInternal> {
```
Comment (coleenp) [inline src/hotspot/share/prims/jvmtiTagMap.hpp:37]: Change to mtServiceability

File: src/hotspot/share/prims/jvmtiTagMap.cpp (hunk 24 above)
Comment (afshin-zafari) [reply src/hotspot/share/prims/jvmtiTagMap.cpp:195]: It isn't used. Removed.

File: src/hotspot/share/prims/jvmtiTagMapTable.cpp (hunk 21 above)
Comment (afshin-zafari) [reply src/hotspot/share/prims/jvmtiTagMapTable.cpp:114]: The methods of HashmapTable are void and have assert to verify the expected function (add/update/remove) is done.

**Review by coleenp — APPROVED** [review]
I have no further comments.  This patch passes testing tier 1-7 for me.

File: src/hotspot/share/prims/jvmtiTagMapTable.cpp [hunk 29]
```diff
@@ -38,189 +38,109 @@
...
-
-JvmtiTagMapEntry* JvmtiTagMapTable::find(int index, unsigned int hash, oop obj) {
-  assert(obj != NULL, "Cannot search for a NULL object");
+void JvmtiTagMapTable::clear() {
+  struct RemoveAll {
+    bool do_entry(const JvmtiTagMapKey& entry, const jlong& tag) {
+      return true;
+    }
+  } remove_all;
+  _table.unlink(&remove_all);
```
Comment (dholmes-ora) [inline src/hotspot/share/prims/jvmtiTagMapTable.cpp:80]: This logic is a bit unclear - can you add some comments. I guess I need to understand what unlink does when reading this code.

File: src/hotspot/share/prims/jvmtiTagMapTable.cpp (hunk 11 above)
Comment (dholmes-ora) [reply src/hotspot/share/prims/jvmtiTagMapTable.cpp:86]: For clarity `clear()` is also called from `JvmtiTagMap::clear()`

File: src/hotspot/share/prims/jvmtiTagMapTable.cpp [hunk 30]
```diff
@@ -38,189 +38,109 @@
...
 
-  WeakHandle w(JvmtiExport::weak_tag_storage(), obj);
-  JvmtiTagMapEntry* p = new_entry(hash, w, tag);
-  Hashtable<WeakHandle, mtServiceability>::add_entry(index, p);
-  ResourceMark rm;
-  log_trace(jvmti, table)("JvmtiTagMap entry added for %s index %d",
-                          obj->print_value_string(), index);
+  JvmtiTagMapKey jtme(obj);
+  jlong* found = _table.get(jtme);
+  return found == NULL ? 0 : *found;
```
Comment (dholmes-ora) [inline src/hotspot/share/prims/jvmtiTagMapTable.cpp:101]: Please use nullptr in new code

File: src/hotspot/share/prims/jvmtiTagMapTable.cpp (hunk 21 above)
Comment (dholmes-ora) [reply src/hotspot/share/prims/jvmtiTagMapTable.cpp:114]: I'm still somewhat concerned about the use of this API and how it will behave in product mode, but okay ...

**Review by dholmes-ora — APPROVED** [review]
Still have a couple of nits regarding comments and some uncertainty about API use/behaviour. But okay. Thanks for the updates.

**Review by dholmes-ora — APPROVED** [review]
